src/control/cmd/daos_server/network.go

```diff
@@ -41,45 +44,51 @@ type networkCmd struct {
 type networkScanCmd struct {
 	cfgCmd
 	logCmd
-	FabricProvider string `short:"p" long:"provider" description:"Filter device list to those that support the given OFI provider (default is the provider specified in daos_server.yml)"`
-	AllProviders   bool   `short:"a" long:"all" description:"Specify 'all' to see all devices on all providers.  Overrides --provider"`
+	FabricProvider string `short:"p" long:"provider" description:"Filter device list to those that support the given OFI provider or 'all' for all available (default is the provider specified in daos_server.yml)"`
 }
 
 func (cmd *networkScanCmd) Execute(args []string) error {
 	var provider string
 
 	if len(args) > 0 {
-		cmd.log.Debugf("An invalid argument was provided: %+v", args)
 		return errors.WithMessage(nil, "failed to execute the fabric and device scan.  An invalid argument was provided.")
 	}
 
-	switch {
-	case cmd.AllProviders:
-		cmd.log.Info("Scanning fabric for all providers")
-	case len(cmd.FabricProvider) > 0:
-		provider = cmd.FabricProvider
-		cmd.log.Infof("Scanning fabric for cmdline specified provider: %s", provider)
-	case len(cmd.config.Fabric.Provider) > 0:
+	provider = cmd.FabricProvider
+	switch provider {
+	case "":
 		provider = cmd.config.Fabric.Provider
-		cmd.log.Infof("Scanning fabric for YML specified provider: %s", provider)
+	case "all":
+		provider = ""
 	default:
-		// all providers case
-		cmd.log.Info("Scanning fabric for all providers")
 	}
 
 	results, err := netdetect.ScanFabric(provider)
 	if err != nil {
 		return errors.WithMessage(err, "failed to execute the fabric and device scan")
 	}
 
-	if provider == "" {
-		provider = "All"
+	interfaces := []*ctlpb.FabricInterface{}
+	for _, fi := range results {
+		interfaces = append(interfaces, &ctlpb.FabricInterface{
+			Provider: fi.Provider,
+			Device:   fi.DeviceName,
+			Numanode: uint32(fi.NUMANode),
+		})
 	}
-	cmd.log.Infof("Fabric scan found %d devices matching the provider spec: %s", len(results), provider)
 
-	for _, sr := range results {
-		cmd.log.Infof("\n%v\n\n", sr)
+	nsr := new(control.NetworkScanResp)
+	hr := control.HostResponse{
+		Addr:    "localhost",
+		Message: &ctlpb.NetworkScanResp{Interfaces: interfaces},
 	}
+	nsr.AddHostResponse(&hr)
```

Comment:
Hmm. I wasn't imagining that you'd need to build up all of the protobuf stuff just to create a `control.HostFabricMap`... I guess the issue is that there's some useful stuff (building deduped providers list, etc.) in `NetworkScanResp.AddHostResponse()`, which I see you've turned into an exported method.

The goal for the control API is to limit leakage of these kinds of details outside of the API, so I don't really want to expose that as part of the API surface.

What I suggest to make this a bit cleaner is to slightly modify `control.HostFabric` to add a `AddInterface(hfi *HostFabricInterface)` method. This would take care of appending it to the Interfaces slice and could also update the Providers slice. It's slightly inefficient to do the Providers update N times rather than once, but this isn't a performance-sensitive area and I think it enables better code. Something like this:

```go
func (hf *HostFabric) AddInterface(hfi *HostFabricInterface) {
        hf.Interfaces = append(hf.Interfaces, hfi)
        hf.Providers = append(hf.Providers, hfi.Provider)
        hf.Providers = dedupeStringSlice(hf.Providers)
} 
```

This way you'd allow API users to work with API types rather than pulling in protobufs stuff which just gets thrown away after converting to native anyhow. Your code would just create a HostFabricMap and add one entry, like so:
```go
hf := &control.HostFabric{}
for _, fi := range results {
        hf.AddInterface(&control.HostFabricInterface{
                Provider: fi.Provider,
                Device:   fi.DeviceName,
                NumaNode: uint32(fi.NUMANode),
        })
}

hfm := make(control.HostFabricMap)
if err := hfm.Add("localhost", hf); err != nil {
        // whoops
}
```

And then you just provide `hfm` to the pretty printer. I know it seems like a bunch of extra work to just print for one host, but I think that the long-term benefit of only having to maintain one pretty-printer implementation makes it worthwhile.

Reply:
Done.  This is nice and clean.